File: isCousins.java
public class isCousins {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    public boolean isCousins(TreeNode root, int x, int y) {
        // return isCousinsBFSWithParentQueue(root, x, y);
        return isCousinsBFSWithoutParentQueue(root, x, y);
        // return isCousinsDFS(root, x, y);
    }

    // TC: O(N) where N is number of nodes
    // SC: O(N) where N is number of nodes
    private boolean isCousinsBFSWithParentQueue(TreeNode root, int x, int y) {
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<TreeNode> parentQueue = new LinkedList<>();
        queue.add(root);
        parentQueue.add(null);
        boolean xFound = false, yFound = false;
        TreeNode xParent = null, yParent = null;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.poll();
                TreeNode currParent = parentQueue.poll();
                if (curr.val == x) {
                    xFound = true;
                    xParent = currParent;
                }
                if (curr.val == y) {
                    yFound = true;
                    yParent = currParent;
                }
                if (curr.left != null) {
                    queue.add(curr.left);
                    parentQueue.add(curr);
                }
                if (curr.right != null) {
                    queue.add(curr.right);
                    parentQueue.add(curr);
                }
            }
            if (xFound && yFound) return xParent != yParent;
            if (xFound || yFound) return false;
        }
        return true;
    }

    int xLevel = 0, yLevel = 0;
    TreeNode xParent = null, yParent = null;
    // TC: O(N) where N is number of nodes
    // SC: O(H) where H is height of tree
    private boolean isCousinsDFS(TreeNode root, int x, int y) {
        dfs(root, x, y, 0, null);
        return xLevel == yLevel && xParent != yParent;
    }

    private void dfs(TreeNode root, int x, int y, int level, TreeNode parent) {
        if (root == null) return;
        if (root.val == x) {
            xLevel = level;
            xParent = parent;
        }
        if (root.val == y) {
            yLevel = level;
            yParent = parent;
        }
        dfs(root.left, x, y, level + 1, root);
        dfs(root.right, x, y, level + 1, root);
    }

    // TC: O(N) where N is number of nodes
    // SC: O(N) where N is number of nodes
    private boolean isCousinsBFSWithoutParentQueue(TreeNode root, int x, int y) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        boolean xFound = false, yFound = false;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.poll();
                if (curr.val == x) {
                    xFound = true;
                }
                if (curr.val == y) {
                    yFound = true;
                }
                if (curr.left != null && curr.right != null) {
                    if (curr.left.val == x && curr.right.val == y) return false;
                    if (curr.left.val == y && curr.right.val == x) return false;
                }
                if (curr.left != null) {
                    queue.add(curr.left);
                }
                if (curr.right != null) {
                    queue.add(curr.right);
                }
            }
            if (xFound && yFound) return true;
            if (xFound || yFound) return false;
        }
        return true;
    }
}

File: isCousins.cpp
// Time Complexity : O(n) where n is the number of nodes
// Space Complexity : O(h) for DFS and O(n) for BFS
// Did this code successfully run on Leetcode : Yes 

// Approach 1: DFS

/*
Two conditions that need to satisfy: 
1. Parents are not the same 
2. Level is the same

In DFS, we can keep track of the level and the return the parent as well
so when we find the node, assign the x_parent and x_level and similar to y

After seeing all the nodes, compare levels and parents and resturn the result
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int x_level, y_level;
    //int works to track parent as the question specifies unique parents
    int x_parent, y_parent;
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if(root == NULL) return false;
        // root, x, y, parent, level
        dfs(root, x, y, NULL, 0);
        // we know at what levels x and y are present and who the parents are
        return (x_level == y_level && x_parent != y_parent);

    }
    void dfs(TreeNode* root, int x, int y, int parent, int level){
        //base
        if(root == NULL) return;

        //logic
        //If we find the node, update the parent and the level
        if(root->val == x){
            x_parent = parent;
            x_level = level;
        }
        if(root->val == y){
            y_parent = parent;
            y_level = level;
        }

        dfs(root->left, x, y, root->val, level+1);
        dfs(root->right, x, y, root->val, level+1);
    }
};


//Approach 2: BFS

/*
We need to satisfy 2 conditions
1. They are in the same level 
2. The nodes are not siblings (which means they are cousins)

To process nodes in the same level, we use a for loop from 0 to the size of the queue at the beginning of each loop
The queue would have all the nodes in each level 

If one node is found, and the other is not -> return false

If both nodes are found, and they are not siblings, we return true

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if(root == NULL) return false;

        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()){
            //this is initialied for every level
            int size = q.size();
            bool x_found = false;
            bool y_found = false;
            for(int i=0; i<size; i++){
                root = q.front();
                q.pop();

                //If we find the node in that level
                if(root->val == x)
                    x_found = true;
                if(root->val == y)
                    y_found = true;

                //if it has both children
                if(root->right && root->left){
                    //we check that it is not the 2 nodes x and y

                    if(root->left->val == x && root->right->val == y)
                        return false;
                    if(root->left->val == y && root->right->val == x)
                        return false;
                }

                //Add the childres to the queue
                if(root->left)
                    q.push(root->left);
                if(root->right)
                    q.push(root->right);
            }
            //after the 'for' loop runs, all the nodes at level n has completed processing
            //now we can check if both the nodes were found
            if (x_found && y_found)
                    return true;

            //we return false even if 1 node was found since the other node is clearly in a different level
            if (x_found || y_found)
                    return false;
        }
        
        return true;
    }
};

